sympy/functions/elementary/trigonometric.py

```diff
@@ -108,21 +108,20 @@ def _peeloff_pi(arg):
 
     """
     pi_coeff = S.Zero
+    rest_terms = S.Zero
     for a in Add.make_args(arg):
-        if a is S.Pi:
-            pi_coeff = S.One
-            break
-        elif a.is_Mul:
-            K = a.coeff(S.Pi)
-            if K and K.is_rational:
-                pi_coeff += K
-    else:
-        if pi_coeff is S.Zero:
-            return arg, S.Zero
+        K = a.coeff(S.Pi)
+        if K and K.is_rational:
+            pi_coeff += K
+        else:
+            rest_terms += a
```

Comment:
It is possible that `rest_terms` can end up being an Add with a large number of terms?

Summing like
```
total = S.Zero
for a in b:
    total += a
```
has `O(N**2)` asymptotic performance because of repeated calls to `Add.flatten`.

If you collect all args and call `Add` once then it is `O(N)`:
```
terms = []
for a in b:
    terms.append(a)
total = Add(*terms)
```

Reply:
This is something new, I was not aware about this. I will make the changes soon.